### PR TITLE
docs: continuous deployment

### PR DESCRIPTION
**Pull request by lonix1**
Continuous deployment page, for newbies like me.

Thanks to @techknowlogick who advised me on [discord](https://discord.com/channels/838698813463724034/838698813463724037/1131383762962370671).

### CODE REVIEW

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+There are various ways to accomplish CD with Woodpecker, depending on your project's specific needs.
+
+## Invoking deploy script via SSH
+
+The final step in your pipeline could SSH into the app server, and run a deployment script.
+
+One of the benefits would be that the deployment script's output could be included in the pipeline's log. However in general, this is a complicated option as it tightly couples the CI and app servers.
+
+An SSH step could be written as a Woodpecker plugin; although there is no official ssh plugin, one does exist for [Drone](https://plugins.drone.io/plugins/ssh) and could be adapted accordingly.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:15]:
```suggestion
An SSH step could be written by using an plugin, like [ssh](https://plugins.drone.io/plugins/ssh) or [git push](https://woodpecker-ci.org/plugins/Git%20Push).
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+On the app server, one should create a script or cron job which polls for asset changes (every minute, say). When a new version is detected, the script redeploys the app.
+
+This option is easy to maintain, but the downside is a short delay (one minute) before new assets are detected.
+
+## Using a configuration management tool
+
+If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool so as to perform the redeployment.
+
+A [Drone](https://plugins.drone.io/plugins/ansible) plugin for Ansible exists and could be adapted accordingly.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:29]:
```suggestion
A plugin for [Ansible](https://plugins.drone.io/plugins/ansible) exists and could be adapted accordingly.
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+A typical CI pipeline contains steps such as: *clone*, *build*, *test*, *package* and *push*. The final build product may be binaries pushed to a git repository or a docker container pushed to a container registry.
+
+When these should be deployed on an app server, the pipeline should include a *deploy* step, which represents the "CD" in CI/CD - the automatic deployment of a pipeline's final product.
+
+There are various ways to accomplish CD with Woodpecker, depending on your project's specific needs.
+
+## Invoking deploy script via SSH
+
+The final step in your pipeline could SSH into the app server, and run a deployment script.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:11]:
```suggestion
The final step in your pipeline could SSH into the app server and run a deployment script.
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+One of the benefits would be that the deployment script's output could be included in the pipeline's log. However in general, this is a complicated option as it tightly couples the CI and app servers.
+
+An SSH step could be written by using an plugin, like [ssh](https://plugins.drone.io/plugins/ssh) or [git push](https://woodpecker-ci.org/plugins/Git%20Push).
+
+## Polling for asset changes
+
+This option completely decouples the CI and app servers, and there is no explicit deploy step in the pipeline.
+
+On the app server, one should create a script or cron job which polls for asset changes (every minute, say). When a new version is detected, the script redeploys the app.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:21]:
```suggestion
On the app server, one should create a script or cron job that polls for asset changes (every minute, say). When a new version is detected, the script redeploys the app.
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+This option is easy to maintain, but the downside is a short delay (one minute) before new assets are detected.
+
+## Using a configuration management tool
+
+If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool so as to perform the redeployment.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:27]:
```suggestion
If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool to perform the redeployment.
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+This option is complex and only suitable in an environment in which you're already using configuration management.
+
+## Using webhooks (recommended)
+
+If your forge (Github, Gitlab, Gitea, etc.) supports webhooks, then you could create a separate listening app that receives a webhook when new assets are available, and redeploys your app.
+
+The listening "app" can be something as simple as a PHP script.
+
+Alternatively, there are a number of popular webhook servers which simplify this process, so you only need to write your actual deployment script. For example, [webhook](https://github.com/adnanh/webhook) and [webhookd](https://github.com/ncarlier/webhookd).
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:39]:
```suggestion
Alternatively, there are a number of popular webhook servers that simplify this process, so you only need to write your actual deployment script. For example, [webhook](https://github.com/adnanh/webhook) and [webhookd](https://github.com/ncarlier/webhookd).
```

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool so as to perform the redeployment.
+
+A plugin for [Ansible](https://plugins.drone.io/plugins/ansible) exists and could be adapted accordingly.
+
+This option is complex and only suitable in an environment in which you're already using configuration management.
+
+## Using webhooks (recommended)
+
+If your forge (Github, Gitlab, Gitea, etc.) supports webhooks, then you could create a separate listening app that receives a webhook when new assets are available, and redeploys your app.
```
Comment (6543) [inline docs/docs/20-usage/33-continuous-deployment.md:35]:
```suggestion
If your forge (Github, Gitlab, Gitea, etc.) supports webhooks, then you could create a separate listening app that receives a webhook when new assets are available and redeploys your app.
```

**Review by 6543 — DISMISSED** [review]

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+## Using a configuration management tool
+
+If you are using a configuration management tool (e.g. Ansible, Chef, Puppet), then you could setup the last pipeline step to call that tool to perform the redeployment.
+
+A plugin for [Ansible](https://plugins.drone.io/plugins/ansible) exists and could be adapted accordingly.
+
+This option is complex and only suitable in an environment in which you're already using configuration management.
+
+## Using webhooks (recommended)
```
Comment (anbraten) [inline docs/docs/20-usage/33-continuous-deployment.md:33]: Not sure what this has to do with WP and why we should recommend it?

File: docs/docs/20-usage/33-continuous-deployment.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# Continuous Deployment
```
Comment (anbraten) [inline docs/docs/20-usage/33-continuous-deployment.md:1]: In general I would like this is a nice overview, but I think its to generic to be actually helpful. I think we could rather add a new category: cookbook #316 or articles to our  new blog with concrete tutorial styled guides. So something like to deploy a docker container with ansible you would first build the container with WP, add it to a registry and then run an ansible script to deploy the new container.

**Comment by 6543** [comment]
I still think we should have a page in docs dedicated to "CD" as I already refer to this pull, while talking to people about woodpecker and if it can do CD too

I agree that the docs structure should be bit restructured

**Comment by 6543** [comment]
after https://github.com/woodpecker-ci/woodpecker/pull/3002 is in I guess this pull could be updated?

**Pull request closed without merging**